### ISSUE TITLE
VP9 codec support, or HEVC + yuva420p format for RGBA support

### ISSUE THREAD

**Issue by BattleAxeVR** [nvpro-samples/vk_video_samples#14]
Hi, I'm trying to get yuva420p pixel format into my game engine (for RGBA videos), and so far the only way I can do it is via VP9 codec since HEVC + alpha encoders are only available on Mac. 

I made some test files for yuva420p using VP9 codec (via FFMPEG), which play back fine in VLC but  assert in this sample due to the missing VK_VIDEO_CODEC_OPERATION_DECODE_VP9_BIT_KHR symbol which isn't in vulkan_beta.h yet. 

I presume this is coming at some point and VK_EXT_video_decode_vp9 define is currently disabled for that reason?

#ifdef VK_EXT_video_decode_vp9
        { VK_VIDEO_CODEC_OPERATION_DECODE_VP9_BIT_KHR, "VP9" },
        
I was wondering if there is an alternative for alpha support (aside from using a separate grayscale video file), and if there is a list of codecs <-> pixel formats that are supported by Vulkan Video and the current beta drivers (and this sample). Thanks for the good work, keep it up.

**Comment by BattleAxeVR** [comment]
Also, just in case it helps anyone, I wrote a quick n dirty hack if you don't want the sample to rush through your video files (while waiting until framerate vs refresh-rate synchronization is implemented in the sample).

 while (true) {
        bool quit = false;

        // process all messages
        MSG msg;
        while (PeekMessage(&msg, nullptr, 0, 0, PM_REMOVE)) {
            if (msg.message == WM_QUIT) {
                quit = true;
                break;
            }

            TranslateMessage(&msg);
            DispatchMessage(&msg);
        }

        if (quit) break;

        acquire_back_buffer();

        double t = timer.get();
        float delta_time_sec = static_cast<float>(t - current_time);
        add_frameProcessor_time(delta_time_sec);

        present_back_buffer();

        current_time = t;

        // WAIT
        float desired_time_sec = 1.0f / 60.0f; // this number should be replaced with the 1/framerate value from each video file, just hardcoded it for now since my videos are 60 FPS

        if (delta_time_sec < desired_time_sec)
        {
            float missing_time_sec = desired_time_sec - delta_time_sec;
            DWORD missing_time_ms = (DWORD)(missing_time_sec * 1000.0f);
            Sleep(missing_time_ms);
        }
    }

**Comment by BattleAxeVR** [comment]
A test clip (.mov file) of HEVC codec with alpha is included in this sample:

https://developer.apple.com/documentation/avfoundation/media_playback_and_selection/using_hevc_video_with_alpha

I can play back the "puppets_with_alpha_hevc.mov" file in MPC-HC, VLC and MPV (not sure if alpha is actually working or set to FF, but it plays back at least). But the file fails in this sample due to some assert from FFMPEG deep inside the opaque/undebuggable dll wrapper used in this sample, so I'm stuck. 

If other media players can play it back, this should should too. I'd be willing to dig into ffmpeg to fix it, but not when I don't have access to the wrapper / glue code inside the prebuilt nv_vkvideo_parser DLL.

I've used FFMPEG as a go-between for a GPU-based video decoder back in 2016 and it was far easier to do that than working in this repo, honestly, even though I was using OpenGL and had to access the DXVA video textures through DX->GL interop. But the way this parser sits right in the middle of Vulkan and FFMPEG is causing a lot of grief, I can't even debug the crashes I get with these test videos.

Please, any help would be greatly appreciated! The whole point of FFMPEG's use of LGPL, from what I understand, is that you need to publish the source code of anything that uses/modifies it. I'm not sure how that relates to nv_vkvideo_parser but anyway, it's certainly gums up the works for a few of us here. We shouldn't need proprietary / closed source DLLs for an open-standards' based video decoder, in my opinion.

**Comment by BattleAxeVR** [comment]
Just re-read your comment in the other issue (about the console window popup which still happens, unfortunately), about VP9/AV1. Sorry, totally forgot that was already on your roadmap.

**Comment by zlatinski** [comment]
> I presume this is coming at some point and VK_EXT_video_decode_vp9 define is currently disabled for that reason?

Sorry, VP9 is not supported yet. The Khronos Video TSG has scheduled to work on this Vulkan spec next year.

**Comment by zlatinski** [comment]
> I can play back the "puppets_with_alpha_hevc.mov" file in MPC-HC, VLC and MPV (not sure if alpha is actually working or set to FF, but it plays back at least).

Thank you, @BattleAxeVR! We'll have a look.

> I've also been considering merely using separate grayscale videos and combining multiple streams into one video file (to maintain synchronization so their frames are all decoded at the same time in parallel), would this be possible?
It would open up higher bit depth (12 or 16) grayscale formats which would be better for 3D motion vectors and depth channel encoding (which I need too). 

I'd propose for you to create two Video decode sessions: one for the regular YCbCr and another for the Lima only. Then you can submit both elementary streams in parallel using two different queues one after another in the same queue using the corresponding video decode sessions. In the first case, you are going to get two completion semaphores from both queue submissions - make sure you submit both semaphores for the graphics composition queue to wait on.

>> So I guess my follow-up question would be: is multi-layer HEVC support on the roadmap or am I better off just manually synchronizing different VK Video decoder instances, 1 per file, and sticking to 1 (grayscale) or 3 (RGB) channels per file.

@BattleAxeVR, let me think about this question. We should be able to support the multi-layer HEVC. However, I'm not sure yet if one would get a composed layer directly from the decoder or if the API would have to use the graphics API to composite the layers. On this, I'll get back to you. In the interim, if you have any additional content that you'd like us to support in addition to the one you've supplied, please post it here.

**Comment by zlatinski** [comment]
@BattleAxeVR, I can reproduce the issue with the content from https://developer.apple.com/documentation/avfoundation/media_playback_and_selection/using_hevc_video_with_alpha. We'll address it with the next release.

**Comment by BattleAxeVR** [comment]
Thanks so much! 

HEVC with alpha decoding would be much better for me than VP9 anyway. Now I just gotta find a way to encode HEVC + alpha using yuva420p on a PC (or equivalent higher bit-depth pixel formats with alpha). Any chance you could use the same fix for the encoding side too? I could use the implementation to do encoding. FFMPEG's libx265 doesn't support yuva420p, but the underlying Nvidia hardware decoder chips surely do since GPU decoding works for those files in other media players (probably via DXVA or NVDEC libs).

For grayscale 10/12/16 bit single-channel formats, the current sample appears to only support RGB packed or planar formats, and bit depth is what's used to determine the format, whereas to support grayscale videos, you'd probably need to use the AVPixelFormat value instead of GetBitDepth to determine the output Vulkan format.

![image](https://user-images.githubusercontent.com/11604039/146269355-c978a6d5-98f7-4a5b-8798-dab55543de20.png)

**Comment by BattleAxeVR** [comment]
For grayscale video formats, if you have a chance to look into those, that would be also great (12 or 16 bit for depth would be ideal):

AV_PIX_FMT_GRAY8

AV_PIX_FMT_GRAY12BE,   ///<        Y        , 12bpp, big-endian
AV_PIX_FMT_GRAY12LE,   ///<        Y        , 12bpp, little-endian
AV_PIX_FMT_GRAY10BE,   ///<        Y        , 10bpp, big-endian
AV_PIX_FMT_GRAY10LE,   ///<        Y        , 10bpp, little-endian

AV_PIX_FMT_GRAY16BE,  ///<        Y        , 16bpp, big-endian
AV_PIX_FMT_GRAY16LE,  ///<        Y        , 16bpp, little-endian

Some of the 64 bit 4:4:4 formats including alpha would also be great to pack in 3D MVs into RGB channels and depth into the alpha channel. I'm using this for fire/smoke volumetric video with Embergen, using DLSS which needs the MVs to resampler/reproject properly.

These two formats in particular would be ideal for some of these use cases:

    AV_PIX_FMT_YUVA444P16BE, ///< planar YUV 4:4:4 64bpp, (1 Cr & Cb sample per 1x1 Y & A samples, big-endian)
    
    AV_PIX_FMT_YUVA444P16LE, ///< planar YUV 4:4:4 64bpp, (1 Cr & Cb sample per 1x1 Y & A samples, little-endian)
    
For RGB or RGBA colour data, I have no interest / need for anything but 420 (444 is a waste of storage / bandwidth / bitrate for colours). But for 3D MVs, I'm not sure how well YUV encodings would work for such linear data, if at all. It's possible that full-res Y -> motion vector magnitude, and half-res CbCr for direction could be OK though, as motion vector directions are probably highly correlated in space. Maybe 420 YUVA for 3D MVs with depth in alpha would work, actually. Who knows, it needs testing.  But the higher bit depth is absolutely necessary for storing depth, no question.

**Comment by zlatinski** [comment]
Hi @BattleAxeVR. It appears that providing support for the LUMA-only channel wouldn't be a very straightforward thing to do.  However, towards the start of next year, we plan to add support for multi-layer HEVC (MV-HEVC). Hopefully, this would work for you.

**Comment by BattleAxeVR** [comment]
I think so, yeah. 

Ideally what would work best for volumetric video, to capture an entire animated point cloud, would be, say, one layer of YUVA video for colour and alpha, either 8-bit or 10-bit per channel, for RGBA, (Alpha is full res, like luma, but UV / chroma being half-res is fine for colour channels), and another YUVA 12- or even 16-bit for 3D motion vectors + depth (in alpha channel). 420 might work for 3D MVs, since luma would effectively encode velocity magnitude and chroma would be its direction, which changes possibly (probably) slower than its magnitude in the average case. 

Finally I could use a third layer in the video file for normals, although it might be best to store them in tangent space and use two channels (for consistency with rest of rendering).  If not I'll just use a three-channel RGB format for model-space normals, probably 8-bit 4:4:4.

**Comment by BattleAxeVR** [comment]
Hi, just wondering if these features / issues are still being worked on or on your list of TODOs. Thanks in advance for any updates you could give, as it will help me make an informed decision about how to proceed in my own planning.

**Comment by ValZapod** [comment]
> mov" file in MPC-HC, VLC and MPV (not sure if alpha is actually working or set to FF, but it plays back at least).

All those are based on ffmpeg, h265 decoder as part of ffmpeg does not support decoding alpha hevc. There are a lo of warnings.

**Comment by zlatinski** [comment]
Hi [BattleAxeVR](https://github.com/BattleAxeVR),
Unfortunately, because of a change in the project's priorities, the MV-HEVC feature has been postponed. 
Please accept our apologies for any inconvenience this has caused!

**Comment by BattleAxeVR** [comment]
Any chance you can take a few minutes to fix popup issue? I reported that a year ago. The rest of the feature requests I can work around just by decoding multiple streams using different instances.

**Comment by zlatinski** [comment]
> Any chance you can take a few minutes to fix popup issue? I reported that a year ago. The rest of the feature requests I can work around just by decoding multiple streams using different instances.

We are preparing a new release this week where in addition to the DLLs, you are going to get static libraries. When using the parser's static libraries, you should be able to build  with whatever Windows options you'd like - console or not.

**Comment by BattleAxeVR** [comment]
Thank you! Really appreciated.

**Comment by ValZapod** [comment]
>those are based on ffmpeg, h265 decoder as part of ffmpeg does not support decoding alpha hevc. There are a lot of warnings.

Just FYI there is now nuh_layer_id patch. https://patchwork.ffmpeg.org/project/ffmpeg/patch/tencent_5F1F4DA9CFDF5521A752B8C9161A4D23A007@qq.com/

**Comment by BattleAxeVR** [comment]
That's very exciting! If I use a version of FFMPEG with this patch integrated, what does that mean, is RGB+A on HEVC supported Vulkan Video now? Or is it just one more piece of the puzzle necessary for it to work. @zlatinski  ?

**Comment by zlatinski** [comment]
> Or is it just one more piece of the puzzle necessary for it to work.

Maybe. Our HW supports multi-layer decode, but we haven't had the time to bring this to the spec and experiment with it.

**Comment by BattleAxeVR** [comment]
In practice, is there a benefit to this in terms of decoding requirements in the hardware? I assume multi-layer decoding requires multiple video decoding pipelines in parallel behind the scenes. I'm just wondering if HEVC RGB+A can be done in a single decoding stream from start to finish, if you get my meaning. Either way, it would save time some manual code to manage my own multiple stream decoding with multiple instances of the vulkan video decoder running in parallel and then having to sync them based on the frame index. That's the part I'd rather not have to do manually as it seems much better to do this at a lower level (hidden to the application I mean).

**Comment by ValZapod** [comment]
>I'm just wondering if HEVC RGB+A can be done in a single decoding stream from start to finish

You do understand that ffmpeg's hwaccel cuda is just acceleration to h265, i.e. software native decoder? Software decoder is still used to preparse SEI stuff, not to mentioned container needs to be parsed.

**Comment by BattleAxeVR** [comment]
> Hi [BattleAxeVR](https://github.com/BattleAxeVR),
> Unfortunately, because of a change in the project's priorities, the MV-HEVC feature has been postponed.

@zlatinski Hi Tony (happy new year btw), I just wanted to double check about this, I was wondering if there was any work done (or still planned) for MV-HEVC or VP9 support for Vulkan Video. 

I just wanted to confirm, either way, even if the answer is no (which is fine) before I start working on my own multi-video-file implementation. 

It could be tricky to sync up the decoding of multiple entire video files perfectly, potentially, but I just don't want to spend my time working on something that Nvidia (or others) already have WIP.

**Comment by BattleAxeVR** [comment]
Update: I re-encoded various text videos using latest FFMPEG and V Video sample here, and all these video files (including apple's .MOV example) decode fine regardless of yuva420 pixel format used in FFMPEG, but the alpha channel isn't extracted from the second layer. That's fine, I can read from a second texture and combine it myself without extra duplicated GPU copies. I just need to know which part of the code I need to hack to grab those extra vulkan images or imageviews so I can hook them up to my descriptor. Any ideas?
